### PR TITLE
EventlogTarget: Support for MaximumKilobytes

### PR DESCRIPTION
**Pull request by ajitpeter**
First PR.  Issue #1673 

---
This change is [Reviewable](https://reviewable.io/reviews/nlog/nlog/1810)

### CODE REVIEW

**Comment by 304NotModified** [comment]
Hey thanks for the PR! Will review it!

File: src/NLog/Targets/EventLogTarget.cs [hunk 1]
```diff
@@ -158,6 +158,11 @@ public int MaxMessageLength
         }
 
         /// <summary>
+        /// Gets or sets the maximum Event log size in kilobytes.
+        /// </summary>
+        public Layout MaxKilobytes { get; set; }
```
Comment (304NotModified) [inline src/NLog/Targets/EventLogTarget.cs]: looks good

File: src/NLog/Targets/EventLogTarget.cs [hunk 2]
```diff
@@ -260,8 +265,17 @@ protected override void Write(LogEventInfo logEvent)
                 category = Convert.ToInt16(this.Category.Render(logEvent), CultureInfo.InvariantCulture);
             }
 
+            long maxKilobytes = 0;
+
+            if(this.MaxKilobytes != null)
+            {
+                maxKilobytes = Convert.ToInt64(this.MaxKilobytes.Render(logEvent), CultureInfo.InvariantCulture);
```
Comment (304NotModified) [inline src/NLog/Targets/EventLogTarget.cs]: maybe it's a bit odd to set this property, when writing every message?

First I was thinking to move it to `InitializeTarget`, but then you have a problem with `GetEventLog(logEvent)`.

Maybe we could cache the value? I was think to move it to `GetEventLog` and maybe creating a method for `isCacheUpToDate `

**Comment by 304NotModified** [comment]
I don't know why the unit tests are failing.

File: src/NLog/Targets/EventLogTarget.cs [hunk 3]
```diff
@@ -260,8 +265,17 @@ protected override void Write(LogEventInfo logEvent)
...
+            }
+
             EventLog eventLog = GetEventLog(logEvent);
 
+            eventLog.MaximumKilobytes = maxKilobytes;
```
Comment (304NotModified) [inline src/NLog/Targets/EventLogTarget.cs]: if `this.MaxKilobytes` is not set, then this will set it to 0. ;)

Therefor the failing unit tests

**Comment by snakefoot** [comment]
Curious why you use  a Layout-object for setting an integer value. Is it some special trick?

**Comment by 304NotModified** [comment]
> Curious why you use a Layout-object for parsing an integer value. Is it some special trick?

Do you mean this line:

` Convert.ToInt64(this.MaxKilobytes.Render(logEvent), CultureInfo.InvariantCulture);`

?

**Comment by snakefoot** [comment]
Yes having an integer property declared as "Layout MaxKilobytes" and then using rendering and parsing to convert it back to an integer.

File: src/NLog/Targets/EventLogTarget.cs (hunk 3 above)
Comment (ajitpeter) [inline src/NLog/Targets/EventLogTarget.cs]: I was wondering why the unit tests on my local enviroment did not fail?? Is there any way to simulate the unit tests on appveyor without creating a pull request??

**Comment by 304NotModified** [comment]
> Yes having an integer property declared as "Layout MaxKilobytes" and then using rendering and parsing to convert it back to an integer.

Well then you could use layout renderers `${var:myvar}`, `${gdc:eventlogmax}` etc.

I have to agree it is not perfect way, but the best way in NLog 4. Related #1234

**Comment by snakefoot** [comment]
> I have to agree it is not perfect way, but the best way in NLog 4

Yes it would make more sense that the correct type was declared for the property, and then the XML-parsing of values was able to recognize Layout-logic and automatically render/parse to the correct property-value-type.

**Comment by 304NotModified** [comment]
> Yes it would make more sense that the correct type was declared for the property, and then the XML-parsing of values was able to recognize Layout-logic and automatically render/parse to the correct property-value-type.

That is possible, but then the property value is not evaluated runtime. 

I was thinking of `Layout<Int>` in the future

**Comment by snakefoot** [comment]
> That is possible, but then the property value is not evaluated runtime.

If the configuration-logic was able to set the correct properties at startup, then the configuration-logic should also be able to set the correct properties on reload (Without using Layout-properties). Would also require some logic regarding perform Reinitialize on a running system.

**Comment by 304NotModified** [comment]
@snakefoot let move the discussion to https://github.com/NLog/NLog/issues/1811, thx! :)

**Comment by 304NotModified** [comment]
@ajitpeter need some help on this?

**Comment by 304NotModified** [comment]
It's now doing nothing ;)

**Comment by ajitpeter** [comment]
@304NotModified  Sorry. These were not intended to be included in the pull request. I was trying out appveyor. I was also trying out what you had mentioned and also debugging simultaneously on my local branch. :-)

File: src/NLog/Targets/EventLogTarget.cs [hunk 4]
```diff
@@ -352,6 +357,7 @@ private EventLog GetEventLog(LogEventInfo logEvent)
             var renderedSource = this.Source != null ? this.Source.Render(logEvent) : null;
             var isCacheUpToDate = eventLogInstance != null && renderedSource == eventLogInstance.Source &&
                                    eventLogInstance.Log == this.Log && eventLogInstance.MachineName == this.MachineName;
+            eventLogInstance.MaximumKilobytes = Convert.ToInt64(this.MaxKilobytes.Render(logEvent));
```
Comment (304NotModified) [inline src/NLog/Targets/EventLogTarget.cs]: This is problematic when the value is empty/0 ;)

**Comment by 304NotModified** [comment]
@ajitpeter needed some help on this?

**Comment by ajitpeter** [comment]
Yes please. I am struggling to replicate the test failure in my local debug environment. Any help would be much appreciated.

**Comment by 304NotModified** [comment]
If all the event log test failing locally , run as admin.    

If the test succeeds locally then it's probably due to information leftovers in the event log. 

Another way is to check the code review and just commit and push.

**Comment by 304NotModified** [comment]
Change looks good. I do prefer auto properties though. 

Please document the 512 (as it's a default of MS - https://msdn.microsoft.com/en-us/library/system.diagnostics.eventlog.maximumkilobytes(v=vs.110).aspx) 

And we need some unit tests.

**Comment by 304NotModified** [comment]
close/repopen for trigger CI

**Comment by 304NotModified** [comment]
There was a stackoverflow in your code ;) 

Have removed that one. Besides that, looks good to me. If it all builds, I will merge it. 

Thanks!

**Comment by ajitpeter** [comment]
Thanks for your help. Much appreciated. :-)

**Comment by 304NotModified** [comment]
Unfortunately the build still fails. Dunno why

**Comment by UgurAldanmaz** [comment]
[From MSDN](https://msdn.microsoft.com/library/system.diagnostics.eventlog.maximumkilobytes(v=vs.110).aspx):

> The MaximumKilobytes property represents the size limit of the event log file. When the event log reaches the size limit, the configured OverflowAction value determines whether new entries are discarded, or whether new entries overwrite older entries.

[MSDN says](https://msdn.microsoft.com/en-us/library/system.diagnostics.eventlog.overflowaction(v=vs.110).aspx) default value of `OverflowAction` is `OverwriteOlder`.

According to test fail messages, logs may are not written. Because they cannot be found when asserting written test log entry.

I suspect if Retention regedit value of the CI machine is `-1`. Check the [OverflowAction source code](https://referencesource.microsoft.com/#System/services/monitoring/system/diagnosticts/EventLogInternal.cs,400).

Tests use the same event log that named as `application`. Application event log is a general instance. So there are maybe so much event log entries that event log file reaches the size limit, then older entries are not overwritten. As a result, unit tests fail.

**Comment by UgurAldanmaz** [comment]
@ajitpeter,

I am not sure at my previous comment. But I suggest you to try. You can write unit tests for your pull request. You just need to supply a unique event log.

* Check the [CreateEventLogTarget](https://github.com/NLog/NLog/blob/ee1d72b6658aa724c6c36f675eb0e1b546aca0cd/tests/NLog.UnitTests/Targets/EventLogTargetTests.cs#L453). You can overload this method to pass log name.
* Call this method with random log name.
* Write your unit tests that use this randomly named event log.

**Comment by 304NotModified** [comment]
Thanks for the info @UgurAldanmaz ,

but maybe i'm missing something. The default is 512 (see [here](https://github.com/NLog/NLog/pull/1810/files#diff-e338ce9c5930f3d84803441ea1c37afeR90) and [here](https://msdn.microsoft.com/en-us/library/system.diagnostics.eventlog.maximumkilobytes(v=vs.110).aspx), and no unit test has been changed, so there is not reason that the unit tests suddenly are reporting errors?

**Comment by 304NotModified** [comment]
I like to merge this, but anyone knows why this is failing the unit tests?

**Comment by ajitpeter** [comment]
I tried but couldn't figure out the unit test failures.

**Comment by 304NotModified** [comment]
it was a nullref, which was swallowed by NLog

fixed in #2114

**Pull request closed without merging**